include/pybind11/detail/eigen_tensor.h

```diff
@@ -0,0 +1,383 @@
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle cast(const Type *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        object parent_object;
+        bool writeable = false;
+        switch (policy) {
+            case return_value_policy::move:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot move from a constant reference");
+                }
+                {
+                    Eigen::aligned_allocator<Type> allocator;
+                    Type *copy = ::new (allocator.allocate(1)) Type(std::move(*src));
+                    src = copy;
+                }
+
+                parent_object = capsule(src, [](void *ptr) {
+                    Eigen::aligned_allocator<Type> allocator;
+                    Type *copy = (Type *) ptr;
+                    copy->~Type();
+                    allocator.deallocate(copy, 1);
+                });
+                writeable = true;
+                break;
+
+            case return_value_policy::take_ownership:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot take ownership of a const reference");
+                }
+                parent_object = capsule(src, [](void *ptr) { delete (Type *) ptr; });
+                writeable = true;
+                break;
+
+            case return_value_policy::copy:
+                parent_object = {};
+                writeable = true;
+                break;
+
+            case return_value_policy::reference:
+                parent_object = none();
+                writeable = !std::is_const<C>::value;
+                break;
+
+            case return_value_policy::reference_internal:
+                // Default should do the right thing
+                parent_object = reinterpret_borrow<object>(parent);
+                writeable = !std::is_const<C>::value;
+                break;
+
+            default:
+                pybind11_fail("pybind11 bug in eigen.h, please file a bug report");
+        }
+
+        handle result = array_t<typename Type::Scalar, compute_array_flag_from_tensor<Type>()>(
+                            H::get_shape(*src), src->data(), parent_object)
+                            .release();
```

Comment:
Do you always want to release()? It seems like you were conditionally keeping it from dec_reffing before. Of course this could also be accomplished by making sure the rvp behavior is correct. Also as a matter of style, the object is usually released on the return statement of the caster.

Reply:
I was conditionally dec_reffing the parent. You always want to release the resulting array. I swapped to your suggested style though, releasing at the return statement is better as it helps avoid bugs.